### PR TITLE
RATEST-231:Failure when validating the vital value of respiratory rate in CaptureVitalsTest

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/RATEST-231

### CODE REVIEW

**Comment by sherrif10** [comment]
Hello @kdaud  , Looking forward for your reviews. This is the fix to the failures on the ci  https://ci.openmrs.org/browse/REFAPP-UI-552/   and https://ci.openmrs.org/browse/CONTRIB-QA-802 .

**Comment by kdaud** [comment]
@sherrif10  do you know why the element was being retrieved before?

**Comment by sherrif10** [comment]
> do you know why the element was being retrieved before?

Which commit  where this element was reverted, I dint think of that in the first but this test was failing

**Comment by kdaud** [comment]
> but this test was failing

Since when ?

And are you able to reproduce it locally ?

**Comment by sherrif10** [comment]
> And are you able to reproduce it locally ?

sure this test was failing according to the workflow, Try to reproduce to confirm, here are the steps
Login
find any Patient
click on any returned patient
click on active visits
click on capture vitals
Enter vitals
Then on entering Respiratory rate, put an integer 110 and press continue, the system is not allowed to enter number biggerthan 110,that's where the problem comes from.

**Comment by sherrif10** [comment]
Hello @kdaud  any more suggestion or approval, we want to fix the ci builds

**Comment by sherrif10** [comment]
cc @kdaud

**Comment by kdaud** [comment]
@sherrif10 do you mind sharing the logs here after reproducing it locally ?

**Comment by sherrif10** [comment]
> do you mind sharing the logs here after reproducing it locally ?

I really have logs minus confirming you that it passes locally, you can follow the steps I indicated above to confirm this workflow

**Comment by kdaud** [comment]
@sherrif10 would be great if you shared the logs here since you are looking into it. However, I will take time & try to reproduce it and give the feedback !

**Comment by sherrif10** [comment]
Will be waiting on your feedback thanks

File: ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java [hunk 1]
```diff
@@ -46,7 +46,8 @@ public void captureVitalsTest() {
         patientCaptureVitalsPage.setWeightField("78");
         patientCaptureVitalsPage.setTemperatureField("36.6");
         patientCaptureVitalsPage.setPulseField("120");
-        patientCaptureVitalsPage.setRespiratoryField("110");
+        patientCaptureVitalsPage.waitForPage();
+        patientCaptureVitalsPage.setRespiratoryField("99");
```
Comment (ibacher) [inline ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java:50]: Why the change in value from 110 -> 99?

File: ui-tests/src/test/java/org/openmrs/reference/page/CaptureVitalsPage.java [hunk 2]
```diff
@@ -21,14 +22,13 @@ public String getPageUrl() {
...
-//        WebElement searchField = findElement(PATIENT_SEARCH);
-//        try {
-//            searchField.sendKeys(text);
-//        } catch(Exception e) {
-//            e.printStackTrace();
-//        }
-//        clickOn(FIRST_FOUND_PATIENT);
-//    }
-
+    public void search(String text) {
```
Comment (ibacher) [inline ui-tests/src/test/java/org/openmrs/reference/page/CaptureVitalsPage.java:25]: Do we need this method at all? Why is this in this PR?

File: ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java [hunk 3]
```diff
@@ -97,10 +98,12 @@ public void setRespiratoryField(String value) {
     }
 
     public void clearPatientBloodPressure1() {
+    	waiter.until(ExpectedConditions.visibilityOfElementLocated(BLOOD_PRESSURE_FIELD_1));
```
Comment (ibacher) [inline ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java:101]:
```suggestion
        waiter.until(ExpectedConditions.visibilityOfElementLocated(BLOOD_PRESSURE_FIELD_1));
```

File: ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java [hunk 4]
```diff
@@ -97,10 +98,12 @@ public void setRespiratoryField(String value) {
...
         findElement(BLOOD_PRESSURE_FIELD_1).clear();
     }
   
     public void clearPatientBloodPressure2() {
+    	waiter.until(ExpectedConditions.visibilityOfElementLocated(BLOOD_PRESSURE_FIELD_2));
```
Comment (ibacher) [inline ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java:106]:
```suggestion
        waiter.until(ExpectedConditions.visibilityOfElementLocated(BLOOD_PRESSURE_FIELD_2));
```

File: ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java [hunk 5]
```diff
@@ -119,7 +122,7 @@ public void setBloodPressureFields(String value1, String value2) {
         } catch (Exception e) {}
     }
 
-    public void clearPatientBloodOxygenSaturation() {
+    public void clearPatientBloodOxygenSaturation() {
```
Comment (ibacher) [inline ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java:125]:
```suggestion
    public void clearPatientBloodOxygenSaturation() {
```

File: ui-tests/src/test/java/org/openmrs/reference/page/CaptureVitalsPage.java (hunk 2 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/CaptureVitalsPage.java:25]: > Do we need this method at all? Why is this in this PR?

I actually think we need this but will investigate what it functions.

File: ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java (hunk 1 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java:50]: > Why the change in value from 110 -> 99?

Maximum value for respiratory rate should be 99.Thats according to the system limitation.

File: ui-tests/src/test/java/org/openmrs/reference/page/CaptureVitalsPage.java (hunk 2 above)
Comment (ibacher) [reply ui-tests/src/test/java/org/openmrs/reference/page/CaptureVitalsPage.java:25]: I don't disagree that we might want it, I'm more of wondering why it's in this PR. It's better to keep a PR to doing the minimum amount necessary to accomplish what the ticket asks. One reason for this is that if this PR causes issues, we can just revert a single commit and fix things whereas if we have multiple things happening in that commit, reverting that commit will unexpectedly break other things.

**Comment by kdaud** [comment]
@sherrif10 am not able to reproduce the issue locally. The test you are trying  to fix (CaptureVitalsTest) is passing locally on all instances and has been passing since it was fixed in May in this [commit.](https://github.com/openmrs/openmrs-distro-referenceapplication/commit/bc7a6b878fb8abfc331bd49ab8d496cab0640a4d)

File: ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java (hunk 1 above)
Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java:50]: @sherrif10 the maximum value of respiratory rate according according to our system is 999 not 99. See the test that validates the maximum value of the parameter [here](https://github.com/openmrs/openmrs-contrib-qaframework/blob/1d2bc4a9d4ccfdff98f49db1efd4f99d882cfabc/src/test/java/org/openmrs/contrib/qaframework/automation/VitalsAndTriagingSteps.java#L122).

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java:50]: > the maximum value of respiratory rate according according to our system is 999 not 99. See the test that validates the maximum value of the parameter [here](https://github.com/openmrs/openmrs-contrib-qaframework/blob/1d2bc4a9d4ccfdff98f49db1efd4f99d882cfabc/src/test/java/org/openmrs/contrib/qaframework/automation/VitalsAndTriagingSteps.java#L122).

Thats what we are trying to resolve , please cross check again in the system

File: ui-tests/src/test/java/org/openmrs/reference/page/CaptureVitalsPage.java (hunk 2 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/CaptureVitalsPage.java:25]: Hey @ibacher . we really dont need this functionality at all.On capturing vitals No where we are searching for patient

File: ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java (hunk 4 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java:106]: Hey @ibacher  , Didnt understand this commit, it might not follow this standard

File: ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java (hunk 1 above)
Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java:50]: For the past, the maximum value has been 999 for Respiratory rate! However, have just noticed a difference in what are servers are reading as per now.

https://demo.openmrs.org/ is reading 999 as was the value before
VF

https://qa-refapp.openmrs.org/openmrs/ is reading 99 which is a new value than what it was reading before
UIx

Is this a change coming from the 2.12 release? Or Its a bug residing in our qa server making now two days old ?

**Comment by sherrif10** [comment]
> am not able to reproduce the issue locally. The test you are trying to fix ([CaptureVitalsTest](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java)) is passing locally on all instances and has been passing since it was fixed in May as per this [commit.](https://github.com/openmrs/openmrs-distro-referenceapplication/commit/bc7a6b878fb8abfc331bd49ab8d496cab0640a4d) !

Which instance are you testing on

**Comment by sherrif10** [comment]
> am not able to reproduce the issue locally

Please before you confirm that things pass or fail, Make sure you are able to reproduce it , Try to test this on a instance or whichever server you are testing on, its better that server to be up to date with current openmrs qa instance.

**Comment by kdaud** [comment]
@sherrif10 I do not think all the changes in this commit are needed, if the issue is coming from the value of the respiratory rate being above the maximum value the system can record, then the change is only needed on the value

**Comment by sherrif10** [comment]
> I do not think all the changes in this commit are needed, if the issue is coming from the value of the respiratory rate being above the maximum value the system can record, then the change is only needed on the value

Sure, we also need to clean the code, Have you confirmed that the respiratory rate is supposed to be 99 and below

**Comment by kdaud** [comment]
> Have you confirmed that the respiratory rate is supposed to be 99 and below

Refer to my report [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/559#discussion_r708579718)

File: ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java (hunk 1 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java:50]: > Is this a change coming from the 2.12 release? Or Its a bug residing in our qa server making now one day old ?

Usually Demo sever contains old functionalities since its a released version, however a server is all time up to date with the current changes in referenceapplication-distro, when a change gets merged, immediately a server gets up to date FYI

**Comment by sherrif10** [comment]
cc @kdaud @ibacher .Looking for your reviews.

File: ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java (hunk 3 above)
Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java:101]: Why adding waiter calls ?  This call is made for elements which take long to be loaded/visible on the page and these fields do not need any waiter calls at all!

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java:101]: > why adding waiter calls ? This call is made for elements which take long to be loaded/visible on the page and these fields do not need any waiter calls at all!

Oh thanks @kdaud , these calls are fixing the issue of visibility of an element brought by the ci, you need to confirm https://ci.openmrs.org/browse/REFAPP-UI-552/

**Comment by kdaud** [comment]
The ticket title and the the commit message needs to be changed to something else because the problem is not with element look up as the ticket and commit states.  The issue is on the vital that is being recorded which is above the maximum value that the system can accept. BTW what is the maximum value of the respiratory rate of a human being ?

If am to approve this PR, I expect a change only at  the value of resperatory rate filed this line code [here](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7f580e7addd1095a0c05bf10dbb470d2a61b190d/ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java#L49) other changes are not really needed.

**Comment by sherrif10** [comment]
> The issue is on the vital that is being recorded which is above the maximum



> The ticket title and the commit message need to be changed to something else because the problem is not with element look up as the ticket and commit states. The issue is on the vital that is being recorded which is above the maximum value that the system can accept. BTW what is the maximum value of the respiratory rate of a human being ?
> 
> If am to approve this PR, I expect a change only at the value of resperatory rate filed this line code [here](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/7f580e7addd1095a0c05bf10dbb470d2a61b190d/ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java#L49) other changes are not really needed.

The vitals are not actually recorded, Am not sure if you really reproduced this test on a instance

> other changes are not really needed.

Did you checkout https://ci.openmrs.org/browse/REFAPP-UI-552/ to confirm why we are adding waiting element please

**Comment by kdaud** [comment]
Bamboo logs provides a hint of the test failure. The only problem is the value being recorded which is above the maximum the system can accept. Once changed to a value within the acceptable range then all will be well, so no need to add waiter calls in the field methods!

**Comment by sherrif10** [comment]
Resolved thanks @kdaud .Any reviews or suggestions.

File: ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java [hunk 6]
```diff
@@ -46,7 +46,8 @@ public void captureVitalsTest() {
         patientCaptureVitalsPage.setWeightField("78");
         patientCaptureVitalsPage.setTemperatureField("36.6");
         patientCaptureVitalsPage.setPulseField("120");
-        patientCaptureVitalsPage.setRespiratoryField("110");
+        patientCaptureVitalsPage.waitForPage();
```
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/CaptureVitalsTest.java:49]: @sherrif10 any reason for adding `waitForPage()` ?

**Comment by kdaud** [comment]
Its worth changing the commit message to something the actually addressing !

**Comment by kdaud** [comment]
@sherrif10 catch my comments above to have this PR approved to fix the UI build failure

**Comment by sherrif10** [comment]
Resolved thanks.

File: ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java (hunk 5 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/PatientCaptureVitalsPage.java]: Resolved thanks

**Review by kdaud — APPROVED** [review]
This commit is worth entering master to resolve [UI](https://ci.openmrs.org/browse/REFAPP-UI) build plan failure.
cc: @dkayiwa
